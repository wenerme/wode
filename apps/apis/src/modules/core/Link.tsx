import React, { AnchorHTMLAttributes, ComponentType } from 'react';
import { Link as ReactRouterLink, LinkProps as ReactRouterLinkProps, useInRouterContext } from 'react-router-dom';
import { default as NextLink, LinkProps as NextLinkProps } from 'next/link';

export const Link: React.FC<
  AnchorHTMLAttributes<HTMLAnchorElement> & {
    to?: string;
    use?: string | ComponentType<any>;
  }
> = ({ children, to, use, ...props }) => {
  props.href ||= to;
  if (typeof window !== 'undefined') {
    if (useInRouterContext()) {
      let p = props as ReactRouterLinkProps;
      p.to = to || props.href || '';
      use = ReactRouterLink;
    }
  }
  if (!use) {
    let p = props as NextLinkProps;
    use ||= NextLink;
  }
  use ||= 'a';

  return React.createElement(use, props, children);
};
