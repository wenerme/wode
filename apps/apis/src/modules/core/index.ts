export { ModuleApp } from './ModuleApp';
export { RootRouteReactor } from './RootRouteReactor';
export { Link } from './Link';
