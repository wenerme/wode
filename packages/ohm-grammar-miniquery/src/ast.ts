import type { ActionDict, IterationNode, MatchResult } from 'ohm-js';
import { MiniQueryGrammar, MiniQuerySemantics } from './grammar';

export type MiniQueryASTNode =
  | {
      type: 'identifier';
      name: string;
    }
  | {
      type: 'logic';
      op: 'and' | 'or';
      a: MiniQueryASTNode;
      b: MiniQueryASTNode;
    }
  | {
      type: 'rel';
      op: 'gt' | 'lt' | 'gte' | 'lte' | 'eq' | 'ne' | 'like' | 'has' | 'in' | 'not in' | 'not like' | 'is' | 'is not';
      a: MiniQueryASTNode;
      b: MiniQueryASTNode;
    }
  | {
      type: 'between';
      op: 'between' | 'not between';
      a: MiniQueryASTNode;
      b: MiniQueryASTNode;
      c: MiniQueryASTNode;
    }
  | {
      type: 'call';
      name: string;
      value: MiniQueryASTNode[];
    }
  | {
      type: 'unary';
      op: 'pos' | 'neg' | 'not';
      value: MiniQueryASTNode;
    }
  | {
      type: 'paren';
      value: MiniQueryASTNode;
    }
  | {
      type: 'array';
      value: MiniQueryASTNode[];
    }
  | {
      type: 'int';
      value: number;
    }
  | {
      type: 'float';
      value: number;
    }
  | {
      type: 'string';
      value: string;
    }
  | {
      type: 'null';
    }
  | {
      type: 'bool';
      value: boolean;
    }
  | {
      type: 'ref';
      name: string[];
    };

const Ops: Record<string, string> = {
  '&&': 'and',
  '||': 'or',
  '=': 'eq',
  '==': 'eq',
  '!=': 'ne',
  '<>': 'ne',
  '>': 'gt',
  '>=': 'gte',
  '<': 'lt',
  '<=': 'lte',
  ':': 'has',
};

export function getMiniQueryASTOp(v: string) {
  const s = v.toLowerCase().trim().replaceAll(/\s+/g, ' ');
  return Ops[s] || s;
}

const actions: ActionDict<any> = {
  nonEmpty(first, _, rest: IterationNode, _pad) {
    return [first].concat(rest.children).map((v) => v.toAST());
  },
  rel(a, op, b) {
    let v = op.sourceString;
    if (op.isIteration()) {
      v = op.children.map((vv) => vv.sourceString).join(' ');
    }
    return {
      type: 'rel',
      op: getMiniQueryASTOp(v) as any,
      a: a.toAST(),
      b: b.toAST(),
      v: 'bool',
    };
  },
  empty() {
    return [];
  },
};

export function toMiniQueryAST(s: string | MatchResult) {
  let match: MatchResult;
  if (typeof s === 'string') {
    match = MiniQueryGrammar.match(s);
  } else {
    match = s;
  }
  if (match.failed()) {
    throw new SyntaxError(`Invalid MiniQuery: ${match.message}`);
  }
  return MiniQuerySemantics(match).toAST();
}

MiniQuerySemantics.addOperation<MiniQueryASTNode>('toAST()', {
  Main(expr, _) {
    return expr.toAST();
  },
  LogicExpr_match(a, op, b) {
    return {
      type: 'logic',
      op: getMiniQueryASTOp(op.sourceString) as any,
      a: a.toAST(),
      b: b.toAST(),
      v: 'bool',
    };
  },
  RelExpr_match: actions.rel,
  RelExpr_match_eq: actions.rel,
  RelExpr_has: actions.rel,
  InExpr_match: actions.rel,
  PredicateExpr_like: actions.rel,
  PredicateExpr_is: actions.rel,
  BetweenExpr_match(a, op, b, _, c) {
    return {
      type: 'between',
      op: op.sourceString as any,
      a: a.toAST(),
      b: b.toAST(),
      c: c.toAST(),
      v: 'bool',
    };
  },
  CallExpr_match(n, _, v, _end) {
    return { type: 'call', name: n.sourceString, value: v.toAST() };
  },
  PriExpr_paren(_, v, _end) {
    return { type: 'paren', value: v.toAST() };
  },
  PriExpr_not(op, v) {
    return { type: 'unary', op: getMiniQueryASTOp(op.sourceString) as any, value: v.toAST(), v: 'bool' };
  },
  PriExpr_pos(op, v) {
    return { type: 'unary', op: getMiniQueryASTOp(op.sourceString) as any, value: v.toAST() };
  },
  PriExpr_neg(op, v) {
    return { type: 'unary', op: getMiniQueryASTOp(op.sourceString) as any, value: v.toAST() };
  },
  Array(_, list, _end) {
    return { type: 'array', value: list.toAST(), v: Array };
  },
  int: (s, _, v) => {
    return { type: 'int', value: parseInt(`${s.sourceString || ''}${v.sourceString}`), v: 'int' };
  },
  float: (i, _, f) => {
    return { type: 'float', value: parseFloat(`${i?.sourceString || 0}.${f.sourceString}`), v: Number };
  },
  string: (_, v, _end) => {
    return { type: 'string', value: v.sourceString, v: 'string' };
  },
  ident: (a, b) => {
    return { type: 'identifier', name: [a, b].map((v) => v.sourceString).join('') };
  },
  null: (_) => {
    return { type: 'null', v: 'null' };
  },
  bool: (v) => {
    return { type: 'bool', value: v.sourceString.toLowerCase() === 'true', v: 'bool' };
  },
  ref(a, b, c: IterationNode) {
    return {
      type: 'ref',
      name: [a.sourceString].concat(
        c.children.map((v) => {
          const ast = v.toAST();
          return ast.name || ast.value;
        }),
      ),
    };
  },
  TrailNonEmptyListOf: actions.nonEmpty,
  EmptyListOf: actions.empty,
});
