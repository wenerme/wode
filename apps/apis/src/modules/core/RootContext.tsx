import React, { type ReactNode, useCallback, useState } from 'react';
import { Setup } from 'common/src/layouts';
import { getSiteConfStore } from 'common/src/system/components';
import { SessionProvider } from 'next-auth/react';
import { useStore } from 'zustand';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ErrorSuspenseBoundary } from '@wener/reaction';
import { createReactClient, trpc } from '../utils/trpc';

export const RootContext: React.FC<{ children?: ReactNode }> = ({ children }) => {
  const [queryClient] = useState(() => new QueryClient({ defaultOptions: { queries: { staleTime: 60 * 5 * 1000 } } }));
  const [trpcClient] = useState(() => createReactClient());
  const { basePath, baseUrl } = useStore(
    getSiteConfStore(),
    useCallback((s) => s.auth, []),
  );
  return (
    <ErrorSuspenseBoundary>
      <trpc.Provider client={trpcClient} queryClient={queryClient}>
        <QueryClientProvider client={queryClient}>
          <SessionProvider basePath={basePath} baseUrl={baseUrl} refetchInterval={5 * 60}>
            <Setup>{children}</Setup>
          </SessionProvider>
        </QueryClientProvider>
      </trpc.Provider>
    </ErrorSuspenseBoundary>
  );
};
