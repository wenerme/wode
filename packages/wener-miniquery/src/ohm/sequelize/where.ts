/* eslint-disable @typescript-eslint/no-non-null-assertion */
import {
  col,
  DataTypes,
  fn,
  Op,
  where,
  type Association,
  type DataType,
  type Includeable,
  type IncludeOptions,
  type Model,
  type ModelStatic,
  type Sequelize,
  type WhereOptions,
} from '@sequelize/core';
import type { MatchResult } from 'ohm-js';
import { toMiniQueryAST, type MiniQueryASTNode } from '../ast';

export interface SequelizeWhereOptions {
  sequelize: Sequelize;
  Model: ModelStatic<Model>;
  where?: WhereOptions[];
  include?: IncludeOptions[];
  // whitelist function calls
  functions?: Record<
    string,
    {
      arity?: number;
    }
  >;
}

type WhereContext = Omit<SequelizeWhereOptions, 'functions'> & {
  current?: any;
  functions: Record<
    string,
    {
      arity?: number;
    }
  >;
  include: Includeable[];
  associations: Association[];
};

export function toSequelizeWhere(
  s: string | MatchResult,
  o: SequelizeWhereOptions,
): { where: WhereOptions; include: IncludeOptions[] } {
  const ctx = { functions: DefaultFunctions, include: [], associations: [], ...o };

  if (!s) {
    return {
      where: {},
      include: [],
    };
  }
  const where = toWhere(toMiniQueryAST(s), ctx);
  ctx.include.push(...ctx.associations);
  return { where, include: ctx.include };
}

const DefaultFunctions: WhereContext['functions'] = {
  date: { arity: 1 },
  length: { arity: 1 },
};

function checkFunctions(name: string, args: any[] = [], o: WhereContext) {
  if (!o.functions[name]) {
    throw new Error(`Invalid function: ${name}`);
  }
}

function isSameType(a: DataType, b: DataType) {
  return a === b || keyOfDataType(a) === keyOfDataType(b);
}

function keyOfDataType(d: DataType) {
  if (typeof d === 'string') {
    return d.toUpperCase();
  }
  return d;
  // if ('key' in d) {
  //   // v7 removed
  //   // return d.key;
  //   // return d.usageContext?.model.name
  //   return d
  // }
  // // fixme
  // return d.name;
}

function resolve(parts: string[], c: WhereContext) {
  const { Model } = c;
  let M = Model;
  const rest = [...parts];

  const all: string[] = [];
  let isAssoc = false;
  let include: IncludeOptions | undefined;

  while (rest.length) {
    const first = rest.shift() as string;
    const attr = M.getAttributes()[first];
    if (attr) {
      if (isSameType(attr.type, DataTypes.JSON) || isSameType(attr.type, DataTypes.JSONB)) {
        all.push(attr.field!);
        all.push(...rest);
        break;
      } else if (rest.length === 0) {
        all.push(attr.field!);
        break;
      } else {
        throw new SyntaxError(`Invalid ref of attr: ${parts.join('.')} type of ${attr.field} is ${attr.type}`);
      }
    } else if (rest.length === parts.length - 1) {
      // first
      // User.createdAt
      if (first === M.name) {
        // todo should add a prefix to prevent ambiguous
        // all.push(first)
        continue;
      }
    }

    const assoc = M.associations[first];
    if (assoc) {
      if (assoc.isMultiAssociation) {
        throw new Error(`Use has for multi association: ${parts.join('.')}`);
      }
      if (!include) {
        include = {
          association: assoc,
          required: true,
        };
        c.include.push(include);
      } else {
        // nested
        include.include = [
          {
            association: assoc,
            required: true,
          },
        ];
      }

      all.push(first);
      M = assoc.target;
      if (!rest.length) {
        throw new SyntaxError(`Invalid ref of an association: ${parts.join('.')}`);
      }
      isAssoc = true;
      continue;
    }

    throw new SyntaxError(`Invalid ref: ${parts.join('.')} of model ${M.name}`);
  }
  if (isAssoc) {
    return `$${all.join('.')}$`;
  }
  return all.join('.');
}

function toWhere(ast: MiniQueryASTNode, o: WhereContext): any {
  let current = o.current ?? {};
  const { Model } = o;

  function attrOfIdentifier(name: string) {
    // can support auto case convert

    const attr = Model.getAttributes()[name];
    if (!attr?.field) {
      throw new Error(`Invalid attribute: ${name}`);
    }
    return attr;
  }

  const setOp = (left: MiniQueryASTNode, op: any, val: any) => {
    if (typeof op === 'string') {
      op = opOf(op);
    }
    let v: any;
    switch (left.type) {
      case 'identifier':
        v = current[attrOfIdentifier(left.name).field!] ??= {};
        break;
      case 'call': {
        const first = left.value[0];
        if (left.value.length !== 1) {
          throw new SyntaxError(`Expected 1 call args, got ${left.value.length}`);
        }
        let cn: string;
        switch (first.type) {
          case 'identifier':
            cn = attrOfIdentifier(first.name).field!;
            break;
          case 'ref':
            // fixme hack replace
            cn = resolve(first.name, o).replaceAll(`$`, '`');
            break;
          default:
            throw new SyntaxError(`Expected identifier or ref for call, got ${first.type}`);
        }

        const name = left.name.toLowerCase();
        checkFunctions(name, [first], o);
        current = where(fn(name, col(cn)), {
          [op]: val,
        });
        return;
      }
      case 'ref':
        // 依然无法 escape `.`
        // https://sequelize.org/docs/v7/core-concepts/model-querying-basics/#querying-json
        // v = _.get(current, left.name);
        // if (!v) {
        //   v = {};
        //   _.set(current, left.name, v);
        // }
        v = current[resolve(left.name, o)] ??= {};
        break;
      default:
        throw new SyntaxError(`Expected identifier, ref or call for left side, got ${left.type}`);
    }
    v[op] = val;
  };

  switch (ast.type) {
    case 'rel':
      {
        setOp(ast.a, ast.op, toWhere(ast.b, { ...o }));
      }
      break;
    case 'between':
      {
        setOp(ast.a, ast.op, [toWhere(ast.b, { ...o }), toWhere(ast.c, { ...o })]);
      }
      break;
    case 'logic':
      {
        const op = opOf(ast.op);
        const c = (current[op] ??= []);
        // collect - flat - optimize
        const a = toWhere(ast.a, { ...o });
        if (a[op]) {
          c.push(...a[op]);
        } else {
          c.push(a);
        }
        const b = toWhere(ast.b, { ...o });
        if (b[op]) {
          c.push(...b[op]);
        } else {
          c.push(b);
        }
      }
      break;

    case 'paren':
      return [toWhere(ast.value, { ...o })];
    case 'int':
    case 'string':
    case 'float':
      return ast.value;
    case 'null':
      return null;
    case 'call': {
      const name = ast.name.toLowerCase();
      checkFunctions(name, ast.value, o);
      return fn(name, ...ast.value.map((v) => toWhere(v, { ...o })));
    }
    case 'identifier': {
      const attr = attrOfIdentifier(ast.name);
      return col(attr.field!);
    }
    default:
      throw new SyntaxError(`Invalid type: ${ast.type}`);
  }
  return current;
}

const OpMap: Record<string, string> = {
  'not between': 'notBetween',
  'not like': 'notLike',
  ilike: 'iLike',
  'not ilike': 'notILike',
  'not in': 'notIn',
  'is not': 'not',
};

function opOf(v: string) {
  const op = Op[(OpMap[v] || v) as 'any'];
  if (!op) {
    throw new SyntaxError(`Unsupported operator: ${v}`);
  }
  return op;
}
