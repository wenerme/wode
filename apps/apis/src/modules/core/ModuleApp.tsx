import React from 'react';
import { type createBrowserRouter } from 'react-router-dom';
import { ThemeProvider } from 'common/src/daisy/theme';
import { type RouteObjects } from 'common/src/router';
import { type DynamicModuleLoader, ModuleProvider, ModuleRouterProvider } from 'common/src/system/module';
import { RootContext } from './RootContext';

export const ModuleApp: React.FC<ModuleAppProps> = ({ loader, createRoutes, createRouter }) => {
  return (
    <ThemeProvider>
      <RootContext>
        <ModuleProvider loader={loader}>
          <ModuleRouterProvider createRoutes={createRoutes} createRouter={createRouter} />
        </ModuleProvider>
      </RootContext>
    </ThemeProvider>
  );
};

export interface ModuleAppProps {
  loader: DynamicModuleLoader;
  createRoutes?: (routes: RouteObjects) => RouteObjects;
  createRouter?: (routes: RouteObjects) => ReturnType<typeof createBrowserRouter>;
}
