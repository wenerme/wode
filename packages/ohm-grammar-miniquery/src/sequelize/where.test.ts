import type { FindOptions, Includeable, ModelStatic, WhereOptions } from '@sequelize/core';
import { DataTypes, Sequelize } from '@sequelize/core';
import { beforeAll, expect, test, TestContext } from 'vitest';
import { toMiniQueryAST } from '../ast';
import { toSequelizeWhere } from './where';

let sequelize: Sequelize;
beforeAll(async () => {
  sequelize = new Sequelize('sqlite::memory:');
  const Profile = await sequelize.define(
    'Profile',
    {
      name: {
        type: DataTypes.STRING,
      },
      attributes: {
        type: DataTypes.JSON,
      },
    },
    {
      underscored: true,
    },
  );
  const User = await sequelize.define(
    'User',
    {
      username: {
        type: DataTypes.STRING,
      },
      name: {
        type: DataTypes.STRING,
      },
      age: {
        type: DataTypes.INTEGER,
      },
      attributes: {
        type: DataTypes.JSON,
      },
      profileId: {
        type: DataTypes.INTEGER,
        references: {
          model: Profile,
          key: 'id',
        },
      },
      avatarUrl: {
        type: DataTypes.STRING,
      },
    },
    {
      underscored: true,
    },
  );
  const Department = await sequelize.define(
    'Department',
    {
      fullName: {
        type: DataTypes.STRING,
      },
      attributes: {
        type: DataTypes.JSON,
      },
    },
    {
      underscored: true,
    },
  );

  const Pet = await sequelize.define(
    'Pet',
    {
      fullName: {
        type: DataTypes.STRING,
      },
      attributes: {
        type: DataTypes.JSON,
      },
    },
    {
      underscored: true,
    },
  );

  const Image = await sequelize.define(
    'Image',
    {
      imageUrl: {
        type: DataTypes.STRING,
      },
      attributes: {
        type: DataTypes.JSON,
      },
    },
    {
      underscored: true,
    },
  );
  Department.hasMany(User, {
    as: 'users',
  });
  User.belongsTo(Department, {
    as: 'department',
  });

  User.hasMany(Pet, {
    as: 'pets',
  });
  Pet.belongsTo(User, {
    as: 'user',
  });

  User.hasOne(Profile, {
    as: 'profile',
  });
  Profile.belongsTo(User, {
    as: 'user',
  });

  Profile.hasOne(Image, {
    as: 'avatar',
  });
  Image.belongsTo(Profile);

  await sequelize.sync();
});
// fixme 无法处理关联
test('sequelize where', async (t) => {
  const { User } = sequelize.models;
  for (const s of [
    '',
    `age > -1`,
    `User.age > -1`,
    // `length(profile.name) > 1`,
    `length(name) > length(username)`,
    `name = 'wener' and age > 18 and age < 80`,
    `name = 'wener' and (age > 18 or age < 80)`,
    `age not between 18 and 80`,
    `date(createdAt) between '2020-01-01' and '2020-01-31' and length(name) > 1 and username is not null and name like 'wen%'`,
    // `attributes.user.name = 'wener'`, // json
    // `attributes.'user name' = 'wener'`, // json
    // `attributes.'user.name' = 'wener'`, // 无法正确 escape
    `age > 0 and age > 0 and age > 0`, // should flatten
    `avatarUrl is not null`, // handle case
  ]) {
    await assertQuery(t, User, s);
  }
});

test('sequelize association', async (t) => {
  const { User } = sequelize.models;
  for (const s of [`profile.name is not null`, `profile.avatar.imageUrl is not null`]) {
    await assertQuery(t, User, s);
  }
});

async function assertQuery(t: TestContext, Model: ModelStatic<any>, query: string) {
  let where: WhereOptions;
  let include: Includeable[];
  try {
    ({ where, include } = toSequelizeWhere(query, {
      sequelize,
      Model,
    }));
  } catch (e) {
    console.log(`MiniQuery: ${query}`);
    console.log(`AST`, toMiniQueryAST(query));
    throw e;
  }

  let sql = '';
  try {
    await Model.findAll({
      logging: (s) => {
        const indexOf = s.indexOf('WHERE ');
        if (indexOf < 0) {
          return (sql = '');
        }
        // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
        return (sql = s.substring(indexOf + 'WHERE '.length, s.length - 1));
      },
      where,
      include,
    });
  } catch (e: any) {
    console.error('ERROR', e?.message);
    console.log(`SQL: ${sql}`);
    console.log(`MiniQuery: ${query}`);
    console.log(`Where`, where);
    throw e;
  }
  expect(sql, `Query: ${query}`).matchSnapshot();
}

test('sequelize incorrect', async (t) => {
  const { User } = sequelize.models;
  for (const s of [
    '1 > 0',
    `name.not_ok = ''`,
    `not_exists = ''`,
    `avatar_url is not null`, // attr not found
    `pets.fullName is not null`, // use has instead
    `pets.fullName: 'wener'`, // not supported yet
  ]) {
    let opt: FindOptions = {};

    expect(() => {
      opt = toSequelizeWhere(s, {
        sequelize,
        Model: User,
      });
      console.log(`Where`, opt.where);
    }, `Query: ${s}`).throw();
    await User.findAll({
      ...opt,
    });
  }
});

test('sequelize type', async (t) => {
  const { type } = sequelize.models.User.getAttributes().attributes as any;
  expect(type).toBeInstanceOf(DataTypes.JSON);
  // sqlite do not support json
  expect(String(type)).toBe('TEXT');
});
