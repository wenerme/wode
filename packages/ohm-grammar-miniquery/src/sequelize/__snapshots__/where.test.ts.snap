// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sequelize association 1`] = `"NOT (\`profile\`.\`name\` IS NULL)"`;

exports[`sequelize association 2`] = `"NOT (\`profile->avatar\`.\`image_url\` IS NULL)"`;

exports[`sequelize where 1`] = `""`;

exports[`sequelize where 2`] = `"\`User\`.\`age\` > -1"`;

exports[`sequelize where 3`] = `"\`User\`.\`age\` > -1"`;

exports[`sequelize where 4`] = `"length(\`name\`) > length(\`username\`)"`;

exports[`sequelize where 5`] = `"\`User\`.\`name\` = 'wener' AND \`User\`.\`age\` > 18 AND \`User\`.\`age\` < 80"`;

exports[`sequelize where 6`] = `"\`User\`.\`name\` = 'wener' AND (\`User\`.\`age\` > 18 OR \`User\`.\`age\` < 80)"`;

exports[`sequelize where 7`] = `"\`User\`.\`age\` NOT BETWEEN 18 AND 80"`;

exports[`sequelize where 8`] = `"(date(\`created_at\`) BETWEEN '2020-01-01' AND '2020-01-31') AND length(\`name\`) > 1 AND NOT (\`User\`.\`username\` IS NULL) AND \`User\`.\`name\` LIKE 'wen%'"`;

exports[`sequelize where 9`] = `"\`User\`.\`age\` > 0 AND \`User\`.\`age\` > 0 AND \`User\`.\`age\` > 0"`;

exports[`sequelize where 10`] = `"NOT (\`User\`.\`avatar_url\` IS NULL)"`;
